Refactorings and work in preparation for PEP 517

pfmoore opened the pull request:
This is the remainder of the work I want to do in preparation for the PEP 517 implementation work. It's probably the last set of changes I can make without user-visible impact, so I'd like to get them merged ahead of the main implementation PR, if possible, as much as anything to simplify that subsequent PR.

Significant changes in this PR (in no particular order):

* I've vendored the `pep517` module. It'll probably need changes as I implement the pip code, so we'll need to revendor at some stage, but this is enough for now.
* I've changed the check for whether people have included setuptools and wheel in their build requirements to be an after-the-fact check that the build environment includes the requirements we need. This makes it easier to check for a recent enough version of setuptools. The precise error people get has probably changed a little, though (there doesn't seem to be a test for this).
* I've added some basic tests, to ensure that the mechanisms are working.
* I've stopped trying to be clever and load `pyproject.toml` on demand. It was too prone to causing bugs. So now you have to call `load_pyproject_toml` before accessing the various attributes based on it (and that load is done in `operations/prepare.py`, in method `prep_for_dist`).

The test suite all passes without modification, which was my main criterion for "can I reasonably consider this as a refactoring/internal reorganisation rather than a functional change?"

I've still not added a `--[no-]use-pep517` command line flag, simply because that would be a user-visible change. I'll add this at a later stage.

Reviews would be appreciated. If I've done this right, it should be a reasonably solid base for adding the PEP 517 implementation, which is going to be the tricky bit - not least because we'll suddenly be switching to the setuptools backend for isolated `setup.py` builds, which switches a chunk of our test suite to a completely new code path...

cjerdonek commented:
Here's one suggestion to simplify things and improve testability. `InstallRequirement` is currently a huge class with a lot of state and functionality (a property that a lot of pip's code base shares). With the current PR, `InstallRequirement.load_pyproject_toml()` both loads `pyproject.toml` _and_ sets the internal state on `InstallRequirement`.

How about moving the implementation of `InstallRequirement.load_pyproject_toml()` to a stand-alone function with no state that takes whatever input arguments it needs and returns an output? Then `InstallRequirement.load_pyproject_toml()` could be simplified to a method a few lines long that simply (1) calls that stand-alone function and then (2) sets its attributes based on the return value. This would make it easier to test the function responsible for loading. Another advantage is that you'd be able to see very easily what arguments the loading function depends on, since it wouldn't be tied to a class with lots of other internal complexity.

pfmoore commented:
@cjerdonek Thanks - that's a good suggestion. I'll look into doing that.

Longer term, the whole of `InstallRequirement` could do with refactoring, but I'm not going to get into that now :-)

pradyunsg commented:
I intend to refractor install requirement and will do so, after you're done with these PEP 517 changes. Until then, I'm mostly going to do minor cleanups.

I'll try take a look at this PR tonight but no promises.

pfmoore commented:
@pradyunsg Cool - no rush over a review, anything you can manage would be appreciated.

pfmoore commented:
Hmm, to refactor `load_pyproject_toml`, I've hit a couple of annoying problems.

1. Messy call signature, because `Installrequirement` has separate properties for the `setup.py` directory, the `setup.py` filename and the `pyproject.toml` filename, and they *don't* just do path joins (there's a bit of bytes/unicode mucking about for Python 2). I'm reluctant to mess around with much in there, as it's fragile and I don't know how well the tests cover it.

2. There's a bug(?) in `InstallRequirement.__str__` - it can return `None` (a non-string value) in some cases if an `InstallRequirement` isn't completely set up. I don't want to get caught up in debugging that, nor do I want to end up just passing the whole requirement object to `load_pyproject_toml`, as that basically defeats the purpose. We only need the `__str__` for error reporting, so it's both annoying and trivial at the same time.

Both of the above are really more to do with general refactoring of `InstallRequirement`, and that's *not* something I want to get sucked into right now. So I think I'm going to hold off on moving `load_pyproject_toml` to a standalone function for now, and just add a comment noting that it's a good candidate for refactoring later.

cjerdonek commented:
> Messy call signature, because Installrequirement has separate properties for the setup.py directory, the setup.py filename and the pyproject.toml filename,

For the bytes/unicode filename code you're referencing, that code is all in the `setup_py` and `pyproject_toml` properties, which you can leave alone. The stand-alone function can simply accept the return value of `self.setup_py` and `self.pyproject_toml` as arguments. Then in the implementation, you would just replace `self.setup_py` and `self.pyproject_toml` with references to those two arguments. For `setup_py_dir`, you can keep that in the method because it's only used once in the last line when setting an attribute on `InstallRequirement`. This way there's nothing for you to mess with.

For the error message referencing `self`, you can handle that by raising that exception in the method itself (either using a try-except around the call to the stand-alone function, or by having it return an error return value). It doesn't look like `InstallRequirement.__str__` depends on `self.use_pep517` being non-None, but it would be easy to check. That's the only attribute the method sets before possibly raising an exception.

There's quite a bit of logic branching inside the `load_pyproject_toml()` method, so it would be nice to have an easy, obvious way to test those code paths / combinations from the get-go.

pfmoore commented:
@cjerdonek Thanks. I see what you mean - I'll think some more about this. I agree with the goal, it's just that once I start down this route, I start to see red and want to hack the whole mess that is `InstallRequirement` apart and clean it up - and the last 2 times I succumbed to that impulse during this exercise, I ended up in a nightmare of obscure test failures and confused commits :smile:

Deep breath, do a little at a time :wink:

pfmoore commented:
OK, after some reflection, I've moved that function:

1. The `__str__` bug turned out to be shallow, so I just fixed that (I don't really want to move error handling into the caller, either we have to use error returns like C, rather than exceptions, or we catch and re-raise an exception, which I don't like).
2. I trimmed down the argument list as you suggested, and it now looks OK to me.

@cjerdonek let me know what you think.

pfmoore commented:
Thanks @pradyunsg - I'm happy to change messages and output, but agree it can be handled in a followup PR just as easily.

I'll leave this a few more days (as I'm likely not to have time to do the next stage for a day or two anyway) and then merge unless anyone has any more review comments.

cjerdonek commented:
Thanks, @pfmoore! I'll give it a look later today.

cjerdonek commented:
The stand-alone function looks great! Thanks a lot for taking the time to do that, @pfmoore.

My only other thought would be to start adding tests for the stand-alone function at some point (to test the various if branches, etc), and also tests for the code paths in `__str__`, including for the bug you fixed, but that can be done later on.

pfmoore commented:
No problem - and thanks for the suggestion in the first place.

Agreed, I really need to look at adding some more tests. But my normal working environment makes it really painful to run tests (a really slow Windows PC, where the proliferation of subprocesses makes things run *really* slowly[1]) so I end up either having to use Travis as my test runner, or wait till I have access to my other PC which is a lot faster. But I will get to it. When I add the actual PEP 517 code path, that will need testing, so I'll probably do all of the tests at that point.

[1] This is something I'm more generally concerned about. To do a pip install of a project from source under PEP 517, can involve a *lot* of subprocesses. One to install the build tools in the build environment, a second to call get_build_requirements_for_wheel, a third to install them, a fourth to call build_wheel. Creating subprocesses on Windows is quite costly, so we could end up with pip being pretty slow on Windows. We've never really looked at performance (to my knowledge) but it might be worth doing so at some point...

cjerdonek commented:
Regarding running tests, what I usually do when I'm preparing a PR is locally to only run the tests I'm specifically working on, e.g. by doing this for a single test module:

    $ tox -e py36 tests/unit/test_vcs.py
    
Or to be even more more specific, by adding a specific test method name--

    $ tox -e py36 tests/unit/test_vcs.py::test_git__get_netloc_and_auth
    
Then I push and let Travis do the full test run after the specific tests work locally.

This workflow has generally been fast enough not to get in the way. The main downside I've found is that the above tox invocation still requires network access. It would be a lot nicer if you could run tests off-line.

benoit-pierre commented:
> I've vendored the pep517 module. It'll probably need changes as I implement the pip code, so we'll need to revendor at some stage, but this is enough for now.

Wouldn't it be better to copy the needed code from the pep517 module instead? Another vendored package is another point of failure when the distributions hobgoblins will inevitably unvendor it. It would make the code/history easier to follow.

Additionally it seems to me that the pep517 module duplicate part of the existing pip code, e.g.: setting up the environment necessary for building a wheel.

pfmoore commented:
> Wouldn't it be better to copy the needed code from the pep517 module instead?

Maybe. The code is (slightly) fiddly, as it runs a script that's in the package (and which uses *other* code from the package) in the subprocess, and getting that right if we cut & paste would take a bit of work.

Pros for vendoring:

* It's easy to do.
* The existing code is tested.
* Longer term we probably want a maintained PEP 517 library anyway (for other tools to use), so duplicating the code now just means we'd either have to maintain 2 copies, or switch to vendoring later. It's similar to the packaging library in that respect.
* Vendoring is our policy for functionality like this.
* Copying has caused us problems elsewhere (appdirs).

Cons for vendoring:

* It pulls in code we're not using.
* Distributions will mess with it and could break stuff.
* I'd immediately look at things I'd prefer to do differently and get sidetracked into changing things rather than doing the PEP 517 implementation work. :smile: :roll_eyes:

IMO, there would have to be a stronger case for copying out the code we needed in order to violate our normal practice of vendoring. But even if it is worth doing, we can do it later - I'd rather vendor for now and copy in later, than do the work of copying in now and later waste that work by switching to vendoring.

cjerdonek commented on src/pip/_internal/pyproject.py:
```diff
@@ -0,0 +1,144 @@
+                "project does not have a setup.py"
+            )
+        use_pep517 = True
+    if build_system and "build-backend" in build_system:
```
This can be an `elif`.

cjerdonek commented on src/pip/_internal/pyproject.py:
```diff
@@ -0,0 +1,144 @@
+    # If we haven't worked out whether to use PEP 517 yet,
+    # and the user hasn't explicitly stated a preference,
+    # we do so if the project has a pyproject.toml file.
+    if use_pep517 is None:
```
This can also be an `elif`.

cjerdonek commented on src/pip/_internal/pyproject.py:
```diff
@@ -0,0 +1,144 @@
+    if use_pep517 is None:
+        use_pep517 = has_pyproject
+
+    if build_system is None and use_pep517:
```
If you move this if-block to be after the:

```python
if not use_pep517:
    return None
```

then you can remove the "`and use_pep517`" from the conditional. I think it's also simpler / clearer this way because `build_system` isn't needed if you're returning with `None`.

pfmoore commented:
@cjerdonek Thanks for the comments. I've addressed them and CI's looking good.
